File: 2461. Maximum Sum of Distinct Subarrays With Length K.py
from typing import *

class Solution:
    def maximumSubarraySum(self, nums: List[int], k: int) -> int:
        n = len(nums)
        if k > n:
            return 0
        
        window = set() # record window' elements
        window_sum = 0
        max_sum = 0
        left = 0

        for right in range(n):
            # check dup: if nums[right] in window, shrink left
            while nums[right] in window:
                window.remove(nums[left])
                window_sum -= nums[left]
                left += 1

            # add current
            window.add(nums[right])
            window_sum += nums[right]

            if right - left + 1 == k:
                max_sum = max(max_sum, window_sum)
                window.remove(nums[left])
                window_sum -= nums[left]
                left += 1
        
        return max_sum

File: 2001-3000/2461. Maximum Sum of Distinct Subarrays With Length K.java
// Medium
// Sliding Window
// Time:O(n),Space:O(k)
// https://leetcode.cn/problems/maximum-sum-of-distinct-subarrays-with-length-k/

import java.util.*;

class Solution {
    public long maximumSubarraySum(int[] nums, int k) {
        int n = nums.length;
        if (n < k) return 0;
        long curSum = 0, maxSum = 0;
        Map<Integer, Integer> map = new HashMap<>();
        int left = 0;

        for (int right = 0; right < n; right++) {
            curSum += nums[right];
            map.put(nums[right], map.getOrDefault(nums[right], 0) + 1);

            // if window size > k
            if (right - left + 1 > k) {
                curSum -= nums[left];
                map.put(nums[left], map.get(nums[left]) - 1);
                if (map.get(nums[left]) == 0) {
                    map.remove(nums[left]);
                }
                left++;
            }
            // deal duplicate element
            while (map.getOrDefault(nums[right], 0) > 1) {
                curSum -= nums[left];
                map.put(nums[left], map.get(nums[left]) - 1);
                if (map.get(nums[left]) == 0) {
                    map.remove(nums[left]);
                }
                left++;
            }
            
            if (right - left + 1 == k) {
                maxSum = Math.max(maxSum, curSum);
            }
        }
        return maxSum;
    }
}
